src/OpenSpaceToolkit/Astrodynamics/Trajectory/Propagator.cpp

```diff
@@ -34,6 +34,10 @@ Propagator::Propagator(const NumericalSolver& aNumericalSolver, const Array<Shar
     : dynamics_(aDynamicsArray),
       numericalSolver_(aNumericalSolver)
 {
+    for (Shared<Dynamics> aDynamics : this->dynamics_)
+    {
+        aDynamics->declareCoordinates(coordinatesBroker_);
+    }
```

Comment:
Instead of storing the various indices in the dynamics objects. Should we return the contribution for each dynamic and then sum it in the propagator? That way the dynamics don't need to keep track of what the individual indices etc. are.

We can update the function signatures so that it accepts a `State` instead of `x` and `dxdt`. That way it can access the relevant values from the `State` object and return a coordinate vector of accelerations or positions or whatever it is (with the relevant coordinate subset).

Reply:
Ok 👍 
See if this would make sense to you/align with what you suggested:

1. Declare coordinates would still exist, and that particular for loop above, would stay as is. At the end of the day, this is a way for each `Dynamics` (or `CoordinatesResolver` for that matter in the future) to state "ey, these are the subsets I play with (either read/write direction)". This happens only once, whenever a `Dynamics` (or `CoordinatesResolver` eventually) is added, and it defines the size of the `State` (so that a dev working at a `Propagator` level doesn't have to worry).

2. `addSubset` function of the `CoordinatesBroker` would no longer return an index and therefore `Dynamics` would no longer store any indexes.

3. Update the `applyCoordinates` signature to use `State`, and map State to/from std::vector inside the `Propagator`, so that it complies with boost's API.

This is aligned with the direction of travel I was working on, which was to:

1. Make the `CoordinatesBroker` a shared pointer (as there will be one per Propagator).

2. Update the current `State` constructor `State(instant, coordinates, frame);` (which currently holds only because there is a hard-coded assumption that coordinates is {pos, vel}) to `State(instant, coordinates, frame, broker);`. At that point in time the `State` instance has all it needs to remain generic: the `coordinates` as well as the `broker` to interpret them.

3. The optimisation is then pushed on to the `State` implementation to make the best use of the `broker` to extract coordinates subsets (for instance, you do not want to fetch for the position index everytime you do `state.getPosition()`) -> but this I will not worry for now.

Reply:
How about this?
1. Dynamics has an Array of CoordinateSubsets, `readCoordinateSubsets` and `writeCoordinateSubsets`. These are hardcoded static values for the class. For example, `AtmosphericDrag` will have a `readCoordinateSubsets` = `{Position, Velocity}` and  `writeCoordinateSubsets` = `{Velocity}`
2. We update `Dynamics` from `applyContribution` to `getContribution`. Where `getContribution` is defined as follows
VectorXd getContribution(const State& aState);
We can access the Position and Velocity from the State object. And the instant as well. (This might not be very performant... but we'll have to see. It's a nice clean interface, but we should make sure our propagation is performant as that is the bread and butter to everything.)
3. We can then put together the contributions in the DynamicalEquations function.
```
void Dynamics::DynamicalEquations(
    const Dynamics::StateVector& x,
    Dynamics::StateVector& dxdt,
    const double& t,
    const Array<Shared<Dynamics>>& aDynamicsArray,
    const Instant& anInstant
)
{
    for (Index i = 0; i < dxdt.size(); ++i)
    {
        dxdt[i] = 0;
    }

    const Instant nextInstant = anInstant + Duration::Seconds(t);
    const State newState = {nextInstant, x, gcrfSPtr_};  // TBI: This could be whatever frame we're integrating in, in the future.

    for (const Shared<Dynamics>& dynamics : aDynamicsArray)
    {
        const VectorXd contributions = dynamics->getContribution(newState);
        for (const CoordinateSet& coordinateSet : dynamics->getWriteContributions)
            Index subsetIndex = CoordinateBroker->getSubsetIndex(coordinateSet);
            for (Index i = 0; i < coordinateSet.getSize(); ++i)
            {
                 dxdt[i + subsetIndex] += contributions[i];
            }
    }
}
```
Something like that, if that makes sense.
Additionally, before running `calculateStatesAt` or `calculateStateAt` we confirm that the State has all the subset coordinates required by the unique set of each dynamics' `readCoordinateSubsets` so that we're sure all the information to propagate exists.

Sorry if this is confusing, github isn't the best place to discuss, we can catch up on this next week!

Reply:
Thanks Viswha, ok I see 👍 
Your approach makes sense and I think it should work.

The only concern I have is performance (as you mentioned).

- For the `readCoordinatesSubsets`, it's fine, it's only called once at the beginning of the propagation.

It's the writing part that I'm more concerned about:

- Every time we do `getContribution(newState)`, each dynamic will have to extract the information that it's interested it from the state. For instance, if it needs position, it's going to do something like 'newState.getPosition()`. Now, that is assuming that there is a `getPosition()` (which exists now, it sort of breaks the 'general state concept' but it's ok as it's too much change to removed `getPosition()` and/or `getVelocity()`. Then what happens inside the `State` instance when `getPosition()` is called? Was it previously stored as a `Vector3d` already, or does it use it's internal `CoordinateBroker` to fetch and extract from the `VectorXd` coordinates? (this is why I was saying that a lot of the optimization would be pushed on to the `State`, nothing wrong with, I just don't know we will come up with something that will at least match the performance of 'not even calling these every time'.
- Every time we do `getWriteContributions` what will happen? Are we returning always a new array or do we create it once and always return the same. If we want to be performant, maybe we created it once, then store it in the dynamics and always return it.
- Every time we do `CoordinateBroker->getSubsetIndex` a map search is performed. 

So, that was my concern. In summary:
- `getContribution(newState)` -> a lot happens here, and it will happen `n x m` times, where `n` is the number of steps and `m` is the number of dynamics.
- `getWriteContributions` -> also happening `n x m` times, but since it always yield the same result (this to me is quite key), we can circumvent by storing once at the beginning and just returning the same value.
- `CoordinateBroker->getSubsetIndex` -> this happens roughly `n x m x p` times where `p` is the number of `write coordinates subsets` each dynamics has. And just like the one above it always returns the same values.

Sorry about the long message. I agree with you maybe this is not the best to discuss, so I'll wait so that we can discuss over a call. But I struggle to see how we are going to beat the 'storing indexes' approach. Maybe not the most beautiful one, but remember indexes stay hidden inside the `Dynamic` implementation. Nobody else knows about it.

I can also update the other 3 dynamics to use indexes and we see how much it takes and measure against the other approach. But on paper it looks like it would at most match it, but I am no Cpp expert so... 🤷

Reply:
I've opened a separate tix for this as it likely involves a discussion on its own : )